src/Compilers/Core/Portable/PEWriter/MetadataVisitor.cs

```diff
@@ -80,11 +80,23 @@ public virtual void Visit(IEventDefinition eventDefinition)
             this.Visit(eventDefinition.GetType(Context));
         }
 
-        public void Visit(IEnumerable<IFieldDefinition> fields)
+        public void Visit(ITypeDefinition containingType, IEnumerable<IFieldDefinition> fields)
         {
-            foreach (IFieldDefinition field in fields)
+            if (containingType.IsValueType)
```

Comment:
>if (containingType.IsValueType) [](start = 12, length = 31)

This feels out of place 
- now we don't call visit for fields and we are doing this for fields that are emitted.
- it also creates code duplication.